cmd/global-query/pkg/distributed/querier.go

```diff
@@ -28,6 +29,14 @@ type Querier interface {
 	CreateQueryWorkload(ctx context.Context, host string, args *query.Args) (*QueryWorkload, error)
 }
 
+// QuerierAnyable extends a "common" Querier with the support to retrieve a list of all hosts / targets
+// available to the Querier
+type QuerierAnyable interface {
+
+	// AllHosts returns a list of all hosts / targets available to the Querier
+	AllHosts() (hosts.Hosts, error)
+}
+
 // APIClientQuerier implements an API-based querier, fulfilling the Querier interface
 type APIClientQuerier struct {
 	apiEndpoints map[string]*client.Config
```

Comment:
You could add

```
hosts.ResolveAnyable
```
to struct to enforce and document that the client querier is able to return all hosts.

Reply:
Isn't this also an issue? When instantiating the global server the resolver & querier are two separate entities (see comment about `initHostListResolver()` and `initQuerier()` above), so now suddenly the querier becomes the resolver (but is never initialized / embedded)?